### PR TITLE
Don't store state from ClientBootstrap.channelInitializer

### LINKED ISSUE #430: gprpc-swift tests stuck (docker)
### New Issue Checklist

- [x] Updated grpc-swift to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/grpc/grpc-swift/blob/master/CONTRIBUTING.md)
- [x] I searched for [existing GitHub issues](https://github.com/grpc/grpc-swift/issues)(especially, there are many [resolved issues related to SSL](https://github.com/grpc/grpc-swift/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+ssl))

### Issue Description

as part of swift-nio-release tools, I run the grpc-swift tests in various configurations. The docker run has been stuck for a loooong time. grpc commit is 283754e80bb30a630e5e06508af654bc1d68fd7d . 

It seems like the promised allocated in  `file="/Users/johannes/.Trash/.nio-release-tools_qP5z8N/grpc-swift/Tests/SwiftGRPCNIOTests/ServerThrowingTests.swift", line=152` is not being fulfilled.

```
(lldb) bt
* thread #1, name = 'SwiftGRPCPackag', stop reason = signal SIGSTOP
  * frame #0: 0x00007f5a2ff41404 libpthread.so.0`__pthread_cond_wait + 196
    frame #1: 0x000055b2b10d32fa SwiftGRPCPackageTests.xctest`ConditionLock.lock(wantedValue=1, self=<unavailable>) at lock.swift:143:23
    frame #2: 0x000055b2b1013454 SwiftGRPCPackageTests.xctest`EventLoopFuture.wait(file="/Users/johannes/.Trash/.nio-release-tools_qP5z8N/grpc-swift/Tests/SwiftGRPCNIOTests/ServerThrowingTests.swift", line=152, self=<unavailable>) at EventLoopFuture.swift:873:14
    frame #3: 0x000055b2b13bba4c SwiftGRPCPackageTests.xctest`implicit closure #1 in ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #4: 0x000055b2b13bbac6 SwiftGRPCPackageTests.xctest`partial apply for implicit closure #1 in ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #5: 0x000055b2b0f66d56 SwiftGRPCPackageTests.xctest`thunk for @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #6: 0x000055b2b13bbb0b SwiftGRPCPackageTests.xctest`thunk for @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #7: 0x00007f5a30240d38 libXCTest.so`closure #1 () -> XCTest.(_XCTAssertionResult in _F1DE58039960C6E2C9E25CD38F016B36) in XCTest.XCTAssertNoThrow<A>(_: @autoclosure () throws -> A, _: @autoclosure () -> Swift.String, file: Swift.StaticString, line: Swift.UInt) -> () + 56
    frame #8: 0x00007f5a30240a58 libXCTest.so`XCTest.XCTAssertNoThrow<A>(_: @autoclosure () throws -> A, _: @autoclosure () -> Swift.String, file: Swift.StaticString, line: Swift.UInt) -> () + 40
    frame #9: 0x000055b2b13bb561 SwiftGRPCPackageTests.xctest`ServerThrowingTests.testBidirectionalStreaming(self=<unavailable>) at ServerThrowingTests.swift:152:5
    frame #10: 0x000055b2b13c1ec9 SwiftGRPCPackageTests.xctest`partial apply for ServerThrowingTests.testBidirectionalStreaming() at <compiler-generated>:0
    frame #11: 0x000055b2b0f66f86 SwiftGRPCPackageTests.xctest`thunk for @escaping @callee_guaranteed () -> (@error @owned Error) at <compiler-generated>:0
    frame #12: 0x000055b2b13c5dab SwiftGRPCPackageTests.xctest`thunk for @escaping @callee_guaranteed () -> (@error @owned Error)partial apply at <compiler-generated>:0
    frame #13: 0x00007f5a30231951 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed () -> (@out (), @error @owned Swift.Error) to @escaping @callee_guaranteed () -> (@error @owned Swift.Error) + 17
    frame #14: 0x00007f5a302316e8 libXCTest.so`partial apply forwarder for closure #1 (XCTest.XCTestCase) throws -> () in XCTest.(test in _3BE257A46ADB477C7BF2D39968B39F9D)<A where A: XCTest.XCTestCase>((A) -> () throws -> ()) -> (XCTest.XCTestCase) throws -> () + 88
    frame #15: 0x00007f5a30231654 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error) + 20
    frame #16: 0x00007f5a302319b9 libXCTest.so`reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) to @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error)partial apply forwarder with unmangled suffix ".37" + 9
    frame #17: 0x00007f5a3021e257 libXCTest.so`partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed XCTest.XCTestCase) -> (@out (), @error @owned Swift.Error) to @escaping @callee_guaranteed (@guaranteed XCTest.XCTestCase) -> (@error @owned Swift.Error) + 39
    frame #18: 0x00007f5a3022f76d libXCTest.so`XCTest.XCTestCase.invokeTest() -> () + 77
    frame #19: 0x00007f5a3022f34c libXCTest.so`XCTest.XCTestCase.perform(XCTest.XCTestRun) -> () + 172
    frame #20: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #21: 0x00007f5a30231dfe libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 254
    frame #22: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #23: 0x00007f5a30231dfe libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 254
    frame #24: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #25: 0x00007f5a30231e20 libXCTest.so`XCTest.XCTestSuite.perform(XCTest.XCTestRun) -> () + 288
    frame #26: 0x00007f5a302353f2 libXCTest.so`XCTest.XCTest.run() -> () + 194
    frame #27: 0x00007f5a3022d9d4 libXCTest.so`XCTest.XCTMain(Swift.Array<(testCaseClass: XCTest.XCTestCase.Type, allTests: Swift.Array<(Swift.String, (XCTest.XCTestCase) throws -> ())>)>) -> Swift.Never + 1652
    frame #28: 0x000055b2b13cc21d SwiftGRPCPackageTests.xctest`main at LinuxMain.swift:10:1
    frame #29: 0x00007f5a2d4f9f45 libc.so.6`__libc_start_main + 245
    frame #30: 0x000055b2b0c95c04 SwiftGRPCPackageTests.xctest`_start + 41
```

package.resolved:

```
{
  "object": {
    "pins": [
      {
        "package": "Commander",
        "repositoryURL": "https://github.com/kylef/Commander.git",
        "state": {
          "branch": null,
          "revision": "e5b50ad7b2e91eeb828393e89b03577b16be7db9",
          "version": "0.8.0"
        }
      },
      {
        "package": "Spectre",
        "repositoryURL": "https://github.com/kylef/Spectre.git",
        "state": {
          "branch": null,
          "revision": "f14ff47f45642aa5703900980b014c2e9394b6e5",
          "version": "0.9.0"
        }
      },
      {
        "package": "swift-nio-ssl-support",
        "repositoryURL": "https://github.com/apple/swift-nio-ssl-support.git",
        "state": {
          "branch": null,
          "revision": "c02eec4e0e6d351cd092938cf44195a8e669f555",
          "version": "1.0.0"
        }
      },
      {
        "package": "swift-nio-zlib-support",
        "repositoryURL": "https://github.com/apple/swift-nio-zlib-support.git",
        "state": {
          "branch": null,
          "revision": "37760e9a52030bb9011972c5213c3350fa9d41fd",
          "version": "1.0.0"
        }
      },
      {
        "package": "SwiftProtobuf",
        "repositoryURL": "https://github.com/apple/swift-protobuf.git",
        "state": {
          "branch": null,
          "revision": "6520fb185db88c0774a929acea1f7d5981a30d3a",
          "version": "1.4.0"
        }
      }
    ]
  },
  "version": 1
}
```

NIO repositories manually pinned to

```
swift-nio: e32a436466b01d4d45555356eb5bbdcdf14b4526
swift-nio-http2: 08db9d8cd4ec5dcb36787f1f7341418cd8c4fa97
swift-nio-ssl: 4f9d64ddd04cd1fe97d74dad7b5a77b882d8bf5d
```

this is Swift 5.0, on docker (ubuntu 14.04)

### LINKED ISSUE #438: NIO client: ClientBootstrap.channelInitializer can be called multiple times
This is just something to watch out for. In NIO 1&2, the `channelInitializer` can be called [multiple times](https://apple.github.io/swift-nio/docs/current/NIO/Classes/ClientBootstrap.html#/s:3NIO15ClientBootstrapC18channelInitializeryACXDAA15EventLoopFutureCyytGAA7Channel_pcF) because we do happy eyeballs with real channels. So you might see one being brought up for IPv4 and one for IPv6. NIO will close the losing one, so as you'd expect only one channel will survive (and be handed to you via `connect`'s future) however there might be multiple `channelInitializer`s run.

The code right in [here](https://github.com/grpc/grpc-swift/blob/bf4bb07d7d7e851136582e7cf697a3bd0ee2df11/Sources/SwiftGRPCNIO/GRPCClientConnection.swift#L43) assumes that this doesn't happen. So worst case (unlikely) is that the `Channel` that NIO decides to kill will complete the `multiplexerPromise` and then you're left in a bad spot: `multiplexerPromise` succeeded with the multiplexer from the closed `Channel`. Likely outcome for this situation would be random hangs. In this [NIO issue](https://github.com/apple/swift-nio/issues/966) I describe how to handle this situation and that we should in the future offer something easier than what you have to do today.

### PR DESCRIPTION
**Pull request by glbrntt**
- channelInitializer can be called multiple times (happy eyeballs)
- the multiplexer was captured but it was never guaranteed to be from
  the "winning" connection
- resolves #438 
- should also resolve #430

### CODE REVIEW

**Comment by glbrntt** [comment]
@MrMage @weissi

**Comment by MrMage** [comment]
I can confirm that this fixes running the tests inside a Docker container. Review forthcoming...

File: Sources/SwiftGRPCNIO/GRPCClientConnection.swift [hunk 1]
```diff
@@ -39,24 +39,22 @@ open class GRPCClientConnection {
...
-        multiplexer.cascade(to: multiplexerPromise)
-        return multiplexer.map { _ in }
+        }.map { _ in }
       }
 
-    return bootstrap.connect(host: host, port: port)
-      .and(multiplexerPromise.futureResult)
-      .map { channel, multiplexer in GRPCClientConnection(channel: channel, multiplexer: multiplexer, host: host, httpProtocol: tlsMode.httpProtocol) }
+    return bootstrap.connect(host: host, port: port).flatMap { channel in
+      channel.pipeline.context(handlerType: HTTP2StreamMultiplexer.self).map { context in
```
Comment (MrMage) [inline Sources/SwiftGRPCNIO/GRPCClientConnection.swift]: What exactly does`context(handlerType:` do?

Comment (glbrntt) [reply Sources/SwiftGRPCNIO/GRPCClientConnection.swift]: Returns the `context` of the first handler of the given type, presumably by walking the pipeline.

Comment (weissi) [reply Sources/SwiftGRPCNIO/GRPCClientConnection.swift]: What George says, but also: https://apple.github.io/swift-nio/docs/current/NIO/Classes/ChannelPipeline.html#/s:3NIO15ChannelPipelineC7context11handlerTypeAA15EventLoopFutureCyAA0B14HandlerContextCGxm_tAA0bJ0RzlF

**Comment by glbrntt** [comment]
More flake from SwiftGRPC 👎 (`testMultipleConnectivityObserversAreCalled`)

**Pull request merged**